lora: thread-safe operations

JordanYates opened the pull request:
Guard the modem against access from multiple contexts while operations are running. This is done to ensure that the modem doesn't end up in undefined states, and that operations that are started complete without interference. This is also required for future asynchronous operations.

Atomics are chosen instead of semaphores or mutex's because it doesn't matter which code path releases ownership of the modem, but we must ensure that only one runs the cleanup code (`Radio.Sleep()`).

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lora: sx12xx_common: sleep after RX timeout

If no packet was received in the provided timeout, manually transition
to sleep mode. If a packet was received, the rx_done callback
automatically transitions the modem to sleep.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lora: fix `lora_recv` parameter description

The timeout parameter for `lora_recv` is no longer a millisecond value
but a standard `k_timeout_t` value.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lora: sx12xx_common: thread-safe API

Ensure that the modem is not asked to perform new operations before the
previous operation completes.

An atomic variable is used instead of a mutex as multiple threads need
to release the lock. A semaphore isn't used as there is no indication
whether `k_sem_give` gives the semaphore or not, which is required to
determine if `Radio.Sleep()` should be run.

`Radio.Sleep()` is only ever run by the context that successfully
releases the modem usage, to guard against double calls.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -80,14 +94,25 @@ int sx12xx_lora_recv(const struct device *dev, uint8_t *data, uint8_t size,
+	}
+
+	/* Clear any previously triggered packets */
+	k_sem_take(&dev_data.data_sem, K_NO_WAIT);
```
This looks a bit suspicious to me. If we can end up in a situation where the semaphore is non-zero here, wouldn't  k_sem_reset be a better option?

JordanYates replied:
This is not entirely correct, given that the semaphore is declared with `K_SEM_MAX_LIMIT`. I actually think it makes more sense to limit this to `1` than change it here.

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -8,18 +8,23 @@
 
 #include "sx12xx_common.h"
 
+#define MODEM_CLAIM(data)       atomic_cas(&data.modem_usage, 0, 1)
```
The commit message specifically mentions a semaphore. Given that the indented use is mutual exclusion, would it make more sense to implement this using vanilla mutexes and set the timeout to K_NO_WAIT?

JordanYates replied:
Mutex's can't be used as multiple calling contexts need to use the object at the same time. Namely the thread that calls the original function (`lora_send`, etc) and the system workqueue, which runs the callbacks.

Using a plain semaphore also doesn't work as you can't tell whether `k_sem_give` (the 'release' operation) succeeded or not.

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -50,24 +55,33 @@ int __sx12xx_configure_pin(const struct device * *dev, const char *controller,
 	dev_data.rssi = rssi;
 	dev_data.snr = snr;
 
+	if (MODEM_FREE(dev_data)) {
```
Would it be better to do this in sx12xx_lora_recv instead? I.e., just setup the buffer here and give the semaphore and call sleep when we take the semaphore in recv. IMO, that would make the code a bit cleaner.

JordanYates replied:
That is an alternative, controlling the state in the callback will still need to be done for the TX case though, so that would make the logic different between TX and RX.

andysan replied:
Good point. Let's keep it symmetric in that case.

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -50,24 +55,33 @@ int __sx12xx_configure_pin(const struct device * *dev, const char *controller,
 static void sx12xx_ev_tx_done(void)
 {
-	Radio.Sleep();
+	if (MODEM_FREE(dev_data)) {
```
We should probably consider making sx12xx_lora_send blocking and enter the sleep state from the send function. That would make the API easier to use since there is currently no way of knowing when the radio is ready short of polling it.

JordanYates replied:
That is the purpose of the follow on PR

andysan replied:
Ack. I saw it just after reviewing this one.

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -80,14 +94,25 @@ int sx12xx_lora_recv(const struct device *dev, uint8_t *data, uint8_t size,
 		LOG_INF("Receive timeout");
 		/* Manually transition to sleep mode on timeout */
-		Radio.Sleep();
+		if (MODEM_FREE(dev_data)) {
```
I'm not sure I understand why you need to check if you can unlock the mutex before entering the sleep state. That seems odd since you know you had exclusive access to the radio until the point you released it.

JordanYates replied:
Because the receive timeout can collide with a packet actually being received, resulting in both code paths trying to call `Radio.Sleep()`, which we can't allow.

andysan replied:
That sounds like a race condition in that case. I think there are cases where the data semaphore, at least in the subsequent changes, can end up in an unexpected state if there is an unfortunate sequence of receive, a timeout, and a send operation. The receive race in this change is _probably_ benign, but I don't like the idea of having code that is racy. It's going to come back and bite us in the rear at some point.

JordanYates replied:
It's racy by definition because we are waiting a fixed duration for an interrupt from an external device. The interrupt can always arrive after we have decided we have timed out, but before we take any action based on that decision. The next update should handle this properly.

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -8,18 +8,23 @@
 
 #include "sx12xx_common.h"
 
+#define MODEM_CLAIM(data)       atomic_cas(&data.modem_usage, 0, 1)
```
Should these be callsed MODEM_ACQUIRE/RELEASE or MODEM_LOCK/UNLOCK instead? I would normally assume that FOO_FREE means that the memory for FOO is deallocated.

JordanYates replied:
Yes

mniestroj commented on drivers/lora/sx12xx_common.c:
```diff
@@ -8,18 +8,23 @@
 #include "sx12xx_common.h"
 
+#define MODEM_ACQUIRE(data) atomic_cas(&data.modem_usage, 0, 1)
+#define MODEM_RELEASE(data) atomic_cas(&data.modem_usage, 1, 0)
```
Personally I would prefer a static function that operates on atomic variable and calls `Radio.Sleep()` implicitly. For example:
```
static inline void modem_release(struct sx12xx_data *data)
{
	if (atomic_cas(&data->modem_usage, 1, 0)) {
		Radio.Sleep();
	}	
}
```
So instead of:
```
	if (MODEM_RELEASE(dev_data)) {
		Radio.Sleep();
	}
```
we could have:
```
	modem_release(dev_data);
```

JordanYates replied:
We would still have the current pattern for the acquire though, as that needs to return.
My preference is to keep both patterns the same.

mniestroj replied:
My way of thinking is that "acquire" can fail if modem is busy, but "release" is an always successful operation. Code that handles LoRa send() and recv() should be only interested in releasing the modem whenever it is no longer used. Power management (calling Radio.Sleep()) is something to be handled in single place, so within `modem_acquire()` and `modem_release()`.

mniestroj replied:
One thing that I noticed now is that with this commit we indeed protect from double `Radio.Sleep()`. But there is also a possible case with:
```
	if (MODEM_RELEASE(dev_data)) {
		Radio.Sleep();
	}
```
that `MODEM_RELEASE()` can be already called and another thread can start `lora_send()` or `lora_recv()` before actually `Radio.Sleep()` will be called. So there is a (not likely, but still) possible race condition here.

andysan replied:
Good point. I didn't spot that particular race. Being unlikely just means it will happen in the field when it's running on a sensor node that has been nailed to a tree top.

andysan reviewed (dismissed):
There are at least two types of race conditions present in this change. As @mniestroj  pointed out, there is a race between releasing the modem and entering the sleep state. This could, incorrectly, disable the modem while it is in the process of sending or receiving a packet. There also seems to be another race in the receive function where the function could potentially be called twice and corrupt the receive buffer. Both of these race classes needs to be fixed before this can be committed.

Commit:
lora: sx12xx_common: transition to k_poll_signal

Transition the receive synchronisation method from a single driver
semaphore to a function specific k_poll_signal. This is required to
allow the modem to be released without introducing race conditions on
the signalling mechanism.

Without this change, the RX can either be signalled before the modem is
released, unblocking the calling thread before the modem is put back to
sleep, or after the modem is released, in which case a second thread
could start using the semaphore before the original thread is signalled.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lora: sx12xx_common: thread safe RX

Make the `lora_recv` operation thread safe by copying memory directly
in the callback instead of deferring copying to the original caller.

To ensure pointer validity, this requires performing operations "inside"
the `modem_release` context.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
Fixing the receive race condition required larger changes than expected, but I believe both conditions have been addressed. @andysan @mniestroj

JordanYates commented:
ping @andysan @mniestroj @Mani-Sadhasivam

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -90,15 +94,42 @@ static bool modem_release(struct sx12xx_data *data)
-	dev_data.snr = snr;
-
-	modem_release(&dev_data);
+	struct k_poll_signal *sig = dev_data.operation_done;
```
I think there is a corner case where you can incorrectly dereference a dangling pointer here. If you are very unlucky, you could potentially have an RX timeout followed by an immediate call to `sx12xx_lora_recv`. If this callback has been pending for some reason (e.g., preemption or lower priority work item), I think this could be called with `operation_done` in state 1. In theory, it seems like you could have a preemption betwen reading the `operation_done` pointer and pre-releasing the modem.

JordanYates replied:
As the comment below explains, I don't think this is possible.

`sig` is only used if the `atomic_cas` succeeds, and if that succeeds we can be sure the calling context is still in `lora_recv`.

andysan reviewed (commented):
Some minor nits.

I can't find any major flaws with the release/acquire logic. It seems like it should work. There may still be a weird corner cases where you end up unexpectedly releasing the modem before a `sx12xx_lora_recv` has finished initialising everything. This type of code sequence would trigger that behaviour:

1. `sx12xx_lora_recv`
2. RX interrupt pending / work item submitted but not serviced
3. Timeout
4. `sx12xx_lora_recv`
5. Old RX interrupt gets serviced while the second `sx12xx_lora_recv` call is running.

This could potentially lead to some surprising behaviour where some parameters aren't updated correctly.  You could potentially end up writing to dangling pointers in this case.

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -78,15 +80,25 @@ int sx12xx_lora_send(const struct device *dev, uint8_t *data,
 int sx12xx_lora_recv(const struct device *dev, uint8_t *data, uint8_t size,
 		     k_timeout_t timeout, int16_t *rssi, int8_t *snr)
 {
+	struct k_poll_signal done;
```
Static is not a solution because if we can't guarantee that done is only ever raised while `sx12xx_lora_recv` is running we also can't guarantee that the output parameters are only written while the function is running. We need to ensure that as those pointers can also be stack variables.

I "think" that the current implementation makes the necessary guarantees,  which I have tried to explain in the `sx12xx_ev_rx_done` callback.

JordanYates replied:
See also the commit message from this change:
> Transition the receive synchronisation method from a single driver
semaphore to a function specific k_poll_signal. This is required to
allow the modem to be released without introducing race conditions on
the signalling mechanism.

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -47,29 +49,72 @@ int __sx12xx_configure_pin(const struct device **dev, const char *controller,
+static bool modem_release(struct sx12xx_data *data)
+{
+	/* Increment atomic so both acquire and release will fail */
+	if (atomic_cas(&data->modem_usage, 1, 2)) {
```
done

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -47,29 +49,72 @@ int __sx12xx_configure_pin(const struct device **dev, const char *controller,
+ */
+static inline bool modem_acquire(struct sx12xx_data *data)
+{
+	return atomic_cas(&data->modem_usage, 0, 1);
```
done

JordanYates commented:
ping @andysan @mniestroj @Mani-Sadhasivam

mniestroj reviewed (approved):
This PR got quite complicated, but I don't see anything obviously wrong here. Definitely an improvement over what we have now, thanks!

Mani-Sadhasivam commented:
Sorry, got a bit swamped by official work. Will get to it in a couple of days.

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -141,9 +141,7 @@ static inline int lora_send(const struct device *dev,
- * @param timeout   Timeout value in milliseconds. API also accepts, 0
-		    for no wait time and SYS_FOREVER_MS for blocking until
-		    data arrives.
+ * @param timeout   Duration to wait for a packet.
```
We should also brief the use case of `K_NO_WAIT` and `K_FOREVER` as done before.

JordanYates replied:
Why is this needed? The behavior is self evident from the arguments.
`K_NO_WAIT` == don't wait at all (will never actually receive anything) 
`K_FOREVER` == wait forever
I don't see other functions documenting the meaning of `K_NO_WAIT` and `K_FOREVER`.

Mani-Sadhasivam replied:
It'd be helpful for the newcomers to Zephyr as they don't need to go through the `k_timeout_t` API doc.

JordanYates replied:
`k_timeout_t` is a core feature of Zephyr. Users should already be familiar with it, and if they're not, going to the struct definition is going to be far more useful to them than giving them two magic values.
From my perspective its equivalent to the following:
```
/**
 * @param a Integer to add.
 *          Adds 2 to b if the value is 2.
 *          Adds 1000 to b if the value is 1000.
 */
int add(int a, int b) { return a + b; };
```
Why specify the values 2 and 1000? Are these the only two values that work? Is the behavior of other integers different?
No, they all work, don't imply they're any different.

Mani-Sadhasivam replied:
No strong feeling here... hence resolving.

Mani-Sadhasivam commented on drivers/lora/sx12xx_common.c:
```diff
@@ -22,14 +22,18 @@
 
+struct sx12xx_rx_params {
+	uint8_t *buf;
+	uint8_t *size;
```
`size` need not be a pointer

JordanYates replied:
It does need to be a pointer, as this is how the caller knows how many bytes were actually received.

Mani-Sadhasivam replied:
You could just return `dev_data.rx_params.size` directly from `sx12xx_lora_recv` instead of `size`. Moreover, we should not try to modify the API arguments without the knowledge of the application.

JordanYates replied:
We cannot just return `dev_data.rx_params.size` directly as the data struct is no longer owned by the caller at that point.
The API arguments are not being modified here. The input variable is still a value, not a pointer.
I'm just using the existing size variable instead of creating an additional temporary.

Mani-Sadhasivam replied:
What I'm trying to say is, you are taking the address of the input argument `size` and storing the actual size in it in case if it differs. So once the `recv` call returns, both `size` argument and the return value will be the same. My argument is, if the API declares a function argument as non-pointer, the API definition should not modify its value.

Mani-Sadhasivam replied:
This can be fixed easily if `dev_data.rx_params.size` is made non-pointer. So it can copy the input `size` value and during the end if `recv`, its value can be returned.

JordanYates replied:
> both size argument and the return value will be the same

That is not the case. Arguments passed to a function by value will never have a different value upon returning from that function. The value is always a copy, as a random reference https://denniskubes.com/2012/08/20/is-c-pass-by-value-or-reference/.

> So it can copy the input size value and during the end if recv, its value can be returned.

My original comment on this was wrong, but not in a way that makes this possible.
As soon as the `sx12xx_ev_rx_done` callback finishes running, the original caller to `lora_recv` no longer has ownership of the `rx_params` struct. Therefore even immediately after returning from `k_poll`, we can make no guarantees about the value of any parameter that it contains. For a value to make its way back to the `lora_recv` caller, it is mandatory for that value to be a pointer to a stack variable which is then populated in the callback itself.

JordanYates replied:
Its no different from
```
int square(int x) {
    x *= x;
    return x;
}

int main(void) {
    int a = 5;
    int sq = square(a);
    printk("%d, %d\n",  a, sq);
}
```
It will print out "5, 25", not "25 25"

Mani-Sadhasivam replied:
doh.. Sorry, my comment about the `size` argument is wrong (Brain fade). But still, please use a separate local variable to copy the `size` value as I still feel that it is not good to take reference of the function argument.

But if you don't wanna do that I'm fine too. Don't want to stretch this discussion too much!

Mani-Sadhasivam commented on drivers/lora/sx12xx_common.c:
```diff
@@ -18,15 +18,15 @@
 
 static struct sx12xx_data {
-	struct k_sem data_sem;
+	struct k_poll_signal *operation_done;
```
Perhaps name this as `rx_sig`? As `operation` is too generic.

JordanYates replied:
The same signal is used in #36577, hence the generic name rather than introducing it here then renaming it in the next PR.

Mani-Sadhasivam replied:
`operation_done` is reused there also. So I don't see any issue in renaming it here and in #36577.

JordanYates replied:
I don't understand. Yes it is reused there. At that point it is used to signal both TX and RX completion. Why do you want it renamed in both PR's?

Mani-Sadhasivam replied:
Ah, didn't see that it is used for both Rx/Tx. Sorry for the confusion.

JordanYates commented:
@cfriedt can this be merged please
